WebDevStudios/wds-blocks
# Add Carousel Block

Closes #79 

### DESCRIPTION ###

Creates a 3-tiered block structure involving 2 new custom blocks, as defined below:

#### Carousel Block ####
- Top-level custom parent block
- Displays child blocks via `InnerBlocks`, limited to the `Slide` Block
- Uses `PreviewToggle` toolbar button to manually switch between "edit" and "preview" modes.
    - Preview: the editor view will mimic the frontend view (i.e., a fully-functional carousel)
    - Edit:  the editor view will display the slides separately in a list so all content is editable

#### Slide Block ####
- Middle-level custom child block
- Limited to displaying only for the `Carousel` Block (i.e., not available as a top-level block or as a child of any other blocks)
- Contains settings specific to the individual slide via inspector panel:
    - Background (color, image, video)
    - Font Color – this sets the default for the current slide, but this can be overridden by child blocks
- Displays child blocks via `InnerBlocks`, limited to certain globally-available blocks: paragraph, heading, buttons
- Defines a simple inner block template, pre-populated with dummy placeholder content (e.g., "Slide Title" for heading block and "Slide Content" for paragraph block)
    - Default template includes heading ("title"), paragraph ("content"), and button

Each block folder contains a basic `README.md` outlining attributes, toolbars, settings, etc.
Filters are included for the following:
- Each block's `InnerBlocks.props`
- Text for previous/next slider buttons
- `glideJS` carousel settings

### SCREENSHOTS ###

Edit Mode:
![screenshot-1](https://user-images.githubusercontent.com/36422618/92284157-b7055b80-eebe-11ea-82ca-dd51efe4f357.png)

Preview Mode:
![screenshot-2](https://user-images.githubusercontent.com/36422618/92284160-b79df200-eebe-11ea-95c7-45312360c9d0.png)

Slide Settings:
![screenshot-1](https://user-images.githubusercontent.com/36422618/92286392-26317e80-eec4-11ea-8ea9-606af1de4c83.png)

### OTHER ###
- [x] Does this issue pass linting? (ESLint)

### STEPS TO VERIFY ###

1. Add `Carousel` block (under WDS Blocks).
2. An initial `Slide` block should appear, with editable title, content, and button.
3. Add at least one more `Slide` block.
4. Edit settings for at least one `Slide` block (details in [`carousel-slide/README.md`](https://github.com/WebDevStudios/wds-blocks/blob/feature/79-carousel-block/src/blocks/carousel-slide/README.md#settings)), confirming changes take immediate effect.
5. Select parent `Carousel` block and toggle between `edit` and `preview` modes.
6. In `preview` mode, confirm navigation buttons and pagination dots are fully-functional.
7. Save and view post, confirming that carousel functions fully and properly on the frontend.

### DOCUMENTATION ###
Will this pull request require updating the WDS Blocks [wiki](https://github.com/WebDevStudios/wds-blocks/wiki)?
Documentation is included in `README.md` files in codebase (further expanded in 2.1).
